fcrepo-exts/migration-utils
# Add micrometer metrics

**JIRA Ticket**: https://jira.lyrasis.org/browse/FCREPO-3692

# What does this Pull Request do?

This PR adds some basic metrics to migration-utils using micrometer. The main goal was to keep everything similar to how the other Fedora 6 metrics are handled, so there are only a few metrics which are actually captured at the moment.

# What's new?

* Added micrometer for capturing metrics
* Add cli option for enabling metrics capture
* Created an `OcflObjectSessionWrapper` to capture metrics on `OcflObjectSessions`
* Created a `CountingInputStream` to capture metrics about bytes processed (based on CountingInputStream from Guava)
* Capture some F3 information
  * Foxml `processObject` and `completeObject`
  * Iterating to the next Foxml FedoraObjectProcessor
* Created a `PrometheusActuator` for publishing metrics for a Promtheus server
* Add sample docker setup for capturing migration utils metrics
  * Now on the wiki: [Migration-Utils-Metrics](https://wiki.lyrasis.org/display/FEDORA6x/Migration+Utils+Metrics)

# How should this be tested?

Start + setup your Prometheus and Grafana instances, then run a migration with the `--enable-metrics` flag and check the results through Grafana.

# Additional Notes:

So a couple things I wanted to highlight:
* The sample duration is happening more frequently than in the normal Fedora 6 config. I did this because the data set I have is processed fairly fast and I needed to actually capture the metrics. Can be tuned further; I ended up using a smaller duration than what's in the commit.
* I provided a sample docker-compose setup for Prometheus + Grafana, which I ended up committing it after the second time deleting the `metrics` directory :). In the README I decided to just link to the wiki, so the setup provided could be moved to the wiki as well if it's too noisy.
* We could pretty easily add an extra cli option for specifying the port to use for the `PrometheusActuator`.

# Interested parties

@fcrepo/committers
